Context/Users/Storage/Context/Users/index.ts

```diff
@@ -37,20 +38,28 @@ export class Users {
 		return result == undefined ? undefined : User.model(this.context, result)
 	}
 	async authenticate(
-		credentials: userwidgets.User.Credentials | userwidgets.User.Key
-	): Promise<userwidgets.User.Key.Creatable | undefined> {
+		credentials: userwidgets.User.Credentials | userwidgets.User.Key,
+		twoFactorCode?: string
+	): Promise<userwidgets.User.Key.Creatable | gracely.Error> {
 		let result: Awaited<ReturnType<Users["authenticate"]>>
 		const current = await this.get()
 		if (!current)
-			result = undefined
+			result = gracely.client.unauthorized()
 		else if (
 			userwidgets.User.Credentials.is(credentials)
 				? !(await Password.verify(credentials.password, current.password, this.context.secret))
 				: credentials.email != current.email
 		)
-			result = undefined
-		else
-			result = userwidgets.User.Key.Creatable.from(User.model(this.context, current))
+			result = gracely.client.unauthorized()
+		else if (current.twoFactor && !twoFactorCode)
+			result = gracely.client.missingHeader("twoFactor", "Time based code generated by authenticator.")
+		else if (current.twoFactor && twoFactorCode && !(await twoFactor.check(current.twoFactor, twoFactorCode)))
+			result = gracely.client.forbidden("Invalid two factor code.")
```

Comment:
forbidden might be the symanticly correct response here but do we want the client to do something here? the ui only pops up the login dialog to retry on a 401. do we want that or should that be pending untill UI is done?

Reply:
The idea is that in this case the UI would reprompt entering of the authenticator code.